Blackbox tests restructuring

paescuj opened the pull request:
- Using `pnpm deploy` to test api in a more realistic environment, getting rid of the current approach https://github.com/directus/directus/blob/5b256f64c689e19ccbef699050320de36ed1f4d5/.github/actions/prepare-test-env/action.yml and the "workaround" (which doesn't really work and thus makes it pointless at the moment anyway)
- Makes it possible again to detect wrong dep types (see https://github.com/directus/directus/pull/18122/commits/29388fb81146381c58fc9ecda742199382796fb8) and wrong location of packages (were hoisted in root until now, see https://github.com/directus/directus/pull/18122/commits/16e5d6cbae625e1e53f7f5f0787c759829b0ee8a#diff-d0ab9c1cc78fba971b5270314ad877706d340d099b42f46a406fc5de6a32fdf0R207)
- Cleaner root & prepares structure for possible  additional tests in future, see https://github.com/directus/directus/pull/15595#issuecomment-1416241566

TODO:
- [x] Make migrations, seeds, extensions paths relative to process so it works no matter from which directory it is called
- [x] Revert https://github.com/directus/directus/pull/18122/commits/15fd1177e429eb481191d2f8167dd41d6dceca83
- [x] Revert testing for all vendors before merging

paescuj committed:
Blackbox tests restructuring

paescuj committed:
Afterfixes

paescuj committed:
TS workaround

paescuj committed:
Approach works ✅ Successfully detected wrong dep type

@directus/types needs to be 'dependencies' instead of 'devDependencies'
because 'Action' enum is used during runtime

paescuj committed:
Small clean-up

paescuj committed:
Adjust paths

paescuj committed:
Fix paths #2

paescuj committed:
Update customEnvironment.ts

paescuj committed:
Merge branch 'main' into blackbox-tests-restructuring

paescuj committed:
Fix api path in schema.test.ts

azrikahar commented on package.json:
```diff
@@ -1,51 +1,21 @@
-		"./packages/*"
-	],
 	"scripts": {
+		"build": "pnpm --recursive --filter '!docs' run build",
```
as we're using `test` instead of `run test` in the "test" script below, should we make them consistent across the board (all with `run`, or all without `run` etc)?

same question for `run` in https://github.com/directus/directus/pull/18122/files#diff-a9d4fab00e8a7db561d841247150ac84dbf3c80ac94f96c4969bc4b66417eb4eR48

paescuj replied:
Good question! For the `run` commands my goal was indeed to make it consistent 👍 For the test scripts I'm relying on the [`pnpm test`](https://pnpm.io/cli/test) command which I think makes it a bit more obvious.

azrikahar commented on api/package.json:
```diff
@@ -89,6 +89,7 @@
 		"@directus/storage-driver-gcs": "workspace:*",
 		"@directus/storage-driver-local": "workspace:*",
 		"@directus/storage-driver-s3": "workspace:*",
+		"@directus/types": "workspace:*",
```
is this move from `devDependencies` to `dependencies` intentional (perhaps to resolve something)?

azrikahar replied:
Answered in the updated PR post!

azrikahar commented on package.json:
```diff
@@ -1,51 +1,21 @@
-		"test:blackbox": "jest tests -c tests-blackbox/jest.config.js",
-		"test:blackbox:watch": "jest tests -c tests-blackbox/jest.config.js --watch",
-		"posttest:blackbox:watch": "ts-node --project ./tests-blackbox/tsconfig.json --transpile-only ./tests-blackbox/setup/teardown.ts"
+    "test": "pnpm --recursive --filter '!tests-blackbox' test"
```
```suggestion
		"test": "pnpm --recursive --filter '!tests-blackbox' test"
```

azrikahar commented on tests/blackbox/logger/redact.test.ts:
```diff
@@ -5,9 +5,10 @@ import { TestLogger } from '@common/test-logger';
 import type { Knex } from 'knex';
+import knex from 'knex';
 import { cloneDeep } from 'lodash';
+import path from 'path';
```
```suggestion
import path from 'node:path';
```

understandably there's still quite a number of places where we aren't using `node:` prefix when importing built in modules, but maybe we should start doing so going forward if that's okay (and maybe update other files with node imports in a another ticket/PR) 👍

We can also leave this for now to do it altogether afterwards!

azrikahar commented on tests/blackbox/setup/setup.ts:
```diff
@@ -1,17 +1,19 @@
-import { clone } from 'lodash';
+import global from './global';
 
+const apiPath = path.join(__dirname, '../../../dist/cli');
```
Maybe this can be combined with https://github.com/directus/directus/pull/18122/files#diff-a82b3abb249145626e26abf82a25f76f0f143492484aaa436a92a7f2ff9981e3R75 as an exported variable from somewhere, but it might also be overkill, so just throwing this thought out there. 😄

paescuj replied:
We were thinking the same 😸👍

azrikahar reviewed (commented):
Apart from the pending TODOs as some tiny review comments, the current changes do make sense 👍

paescuj committed:
Fix indent

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

licitdev committed:
Merge branch 'main' into blackbox-tests-restructuring

paescuj committed:
Use 'node:path' import style

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

paescuj committed:
Revert "TS workaround"

This reverts commit 15fd1177e429eb481191d2f8167dd41d6dceca83.

licitdev commented on package.json:
```diff
@@ -1,51 +1,21 @@
-		"test:blackbox": "jest tests -c tests-blackbox/jest.config.js",
-		"test:blackbox:watch": "jest tests -c tests-blackbox/jest.config.js --watch",
-		"posttest:blackbox:watch": "ts-node --project ./tests-blackbox/tsconfig.json --transpile-only ./tests-blackbox/setup/teardown.ts"
+		"test": "pnpm --recursive --filter '!tests-blackbox' test"
```
```suggestion
		"test": "pnpm --recursive --filter '!tests-blackbox' test",
		"blackbox": "pnpm --filter directus deploy --prod dist && pnpm --filter tests-blackbox test"
```

Can we add a top level blackbox command to simplify the running of blackbox testing?

paescuj committed:
Add top level blackbox script

Co-authored-by: ian <licitdev@gmail.com>

paescuj committed:
Define paths in config.ts, reuse from there & specify cwd

paescuj committed:
Merge branch 'main' into blackbox-tests-restructuring

paescuj committed:
Revert "Approach works ✅ Successfully detected wrong dep type"

This reverts commit 29388fb81146381c58fc9ecda742199382796fb8.

paescuj committed:
Polishing ✨

paescuj committed:
Additional clean-up & update readme a bit

rijkvanzanten commented on packages/composables/tsconfig.json:
```diff
@@ -2,8 +2,7 @@
 	"compilerOptions": {
 		"outDir": "dist",
-		"lib": ["es2022", "DOM"],
-		"ignoreDeprecations": "5.0"
```
Hey hey typescript 5 here we gooooooo

rijkvanzanten reviewed (approved):
Code changes LGTM!

licitdev commented on tests/blackbox/readme.md:
```diff
@@ -22,15 +26,15 @@ The test suite will spin up a fresh copy of the Directus API from the current bu
 
-### Watching for (test) changes
+### Watching for test changes
 
-Use `npm run test:blackbox:watch` to enable Jest's `--watch` mode, especially useful in combination with the flags
+Use `pnpm run test:watch` to enable Jest's `--watch` mode, especially useful in combination with the flags
 above.
 
 This _does not_ watch changes to Directus; it only watches changes to the tests.
```
Let's remove this block, the `watch` option has been removed.

rijkvanzanten replied:
> Should we also remove the script from `package.json`?

As long as it's consistent everywhere 👍🏻  I think we still have some `test:watch` commands, so lets make sure we clean those up together with this. It's easy enough to do `pnpm test --watch` seeing that `pnpm` passes the flags along to the underlying command 🙂

licitdev replied:
Yes, please do! 👍

rijkvanzanten replied:
> As we're already touching those lines (😄), should we remove the test:coverage scripts (and @vitest/coverage-c8 packages?) as well since we're not making use of it yet?

Lets remove the pnpm shorthands, but leave the dev-dep. I actively use it during development to see if I'm "done" by confirming my logic has 100% coverage 👍🏻

paescuj replied:
Done 👍 Ready from my side 😃

paescuj replied:
Should be possible with those commands:
```bash
pnpm -F api run test --coverage <specificTest>
pnpm -F api test -- --coverage <specificTest>
```
Would you still prefer to reintroduce the script?

rijkvanzanten replied:
What I tend to do is `cd` into the `api` folder, and then run `p run test --coverage`. No -F required if you're already in the correct folder 🙂

paescuj committed:
Remove unpopular 'test:{watch,coverage}' scripts

Can be achieved by simply running `pnpm test --watch` resp. `pnpm test
--coverage`

paescuj committed:
Revert testing for all vendors

rijkvanzanten committed:
Merge branch 'main' into blackbox-tests-restructuring

licitdev reviewed (approved):
@paescuj Nice restructure to use `pnpm deploy`! 👍
